data_structures.md

```diff
@@ -11,12 +11,601 @@ kernelspec:
+and on which edge (equivalently, above which node) they occurred.
+The positions are recorded in the {class}`SiteTable`:
+
+```{code-cell} ipython3
+ts.tables.sites
+```
+
+As with node tables, the ``id`` column is **not** actually recorded, but is
+implied by the position in the table.  The results of the
+actual mutations are then recorded:
+
+
+```{code-cell} ipython3
+ts.tables.mutations
+```
+
+This then results in the following (two-locus) haplotypes for the three
+samples:
+
+```{code-cell} ipython3
+for sample, h in enumerate(ts.haplotypes()):
+    print("Sample {sample}: {h}")
+```
+
+
+To create these tables, and the corresponding tree sequence, we would
+create a {class}`TableCollection`, and then use its
+{meth}`~TableCollection.tree_sequence` method:
+
+```{code-cell} ipython3
+tables = tskit.TableCollection(sequence_length=1.0)
+
+# Nodes
+sv = [True, True, True, False, False, False, False]
+tv = [0.0, 0.0, 0.0, 0.4, 0.5, 0.7, 1.0]
+
+for is_sample, t in zip(sv, tv):
+ flags = tskit.NODE_IS_SAMPLE if is_sample else 0
+ tables.nodes.add_row(flags=flags, time=t)
+
+# Edges
+lv = [0.2, 0.2, 0.0, 0.0, 0.8, 0.2, 0.8, 0.8, 0.0, 0.0]
+rv = [0.8, 0.8, 1.0, 0.2, 1.0, 0.8, 1.0, 1.0, 0.2, 0.2]
+pv = [3, 3, 4, 4, 4, 4, 5, 5, 6, 6]
+cv = [0, 2, 1, 2, 2, 3, 0, 4, 0, 4]
+
+for l, r, p, c in zip(lv, rv, pv, cv):
+    tables.edges.add_row(left=l, right=r, parent=p, child=c)
+
+# Sites
+for p, a in zip([0.1, 0.5], ['0', '0']):
+    tables.sites.add_row(position=p, ancestral_state=a)
+
+# Mutations
+for s, n, d in zip([0, 1, 1], [4, 3, 2], ['1', '1', '0']):
+    tables.mutations.add_row(site=s, node=n, derived_state=d)
+```
+
+We can then finally obtain the tree sequence:
+
+```{code-cell} ipython3
+ts = tables.tree_sequence()
+SVG(ts.draw_svg())
+```
+
+:::{todo}
+Point to the metadata tutorial to discuss the metadata columns.
+:::
+
+
+(sec_data_structures_editing)=
+
+## Editing tree sequences
+
+Sometimes we wish to make some minor modifications to a tree sequence that has
+been generated by a simulation. However, tree sequence objects are **immutable**
+and so we cannot edit them in place. To modify a tree sequence, we need to
+extract the underlying {ref}`tables <tskit:sec_table_definitions>` of information, edit these tables,
+and then create a new tree sequence from them.
+These tables succinctly store everything we need to know
+about a tree sequence, and can be manipulated using the {ref}`tskit:sec_tables_api`.
+In the following example, we use this approach
+to remove all singleton sites from a given tree sequence.
+
+```{code-cell} ipython3
+def strip_singletons(ts):
+    tables = ts.dump_tables()
+    tables.sites.clear()
+    tables.mutations.clear()
+    for tree in ts.trees():
+        for site in tree.sites():
+            assert len(site.mutations) == 1  # Only supports infinite sites muts.
+            mut = site.mutations[0]
+            if tree.num_samples(mut.node) > 1:
+                site_id = tables.sites.add_row(
```

Comment:
Better to update here to use the new append (I think it's called?) method. Will be a while before this is updated again, and manually specifying the cols here like this (and leaving out a bunch of them) is a major anti-pattern.

Reply:
I've done this, but doing it for mutations means using `dataclass.replace`. See comments at the end of this PR.